libmamba/src/download/mirror.cpp

```diff
@@ -0,0 +1,169 @@
+    }
+
+    bool operator==(const MirrorID& lhs, const MirrorID& rhs)
+    {
+        return lhs.m_value == rhs.m_value;
+    }
+
+    /*****************
+     * MirrorRequest *
+     *****************/
+
+    MirrorRequest::MirrorRequest(
+        const RequestBase& base,
+        std::string_view lurl,
+        std::optional<header_list> lheaders
+    )
+        : RequestBase(base)
+        , url(lurl)
+        , headers(std::move(lheaders))
+    {
+    }
+
+    /*************************
+     * Mirror implementation *
+     *************************/
+
+    Mirror::Mirror(MirrorID id, std::size_t max_retries)
+        : m_id(std::move(id))
+        , m_max_retries(max_retries)
+    {
+    }
+
+    const MirrorID& Mirror::id() const
+    {
+        return m_id;
+    }
+
+    auto Mirror::get_request_generators() const -> request_generator_list
+    {
+        return get_request_generators_impl();
+    }
+
+    std::size_t Mirror::max_retries() const
+    {
+        return m_max_retries;
+    }
+
+    std::size_t Mirror::successful_transfers() const
+    {
+        return m_successful_transfers;
+    }
+
+    std::size_t Mirror::failed_transfers() const
+    {
+        return m_failed_transfers;
+    }
+
+    bool Mirror::can_accept_more_connections() const
+    {
+        return !m_allowed_connections.has_value() || m_running_transfers < m_allowed_connections;
+    }
+
+    bool Mirror::can_retry_with_fewer_connections() const
+    {
+        return m_running_transfers > 0
+               || (m_successful_transfers > 0 && m_failed_transfers < m_max_tried_connections);
+    }
+
+    namespace
+    {
+        using lock_guard = std::lock_guard<std::mutex>;
+    }
+
+    void Mirror::cap_allowed_connections()
+    {
+        lock_guard lock(m_stats_mutex);
+        if (m_running_transfers > 0)
+        {
+            m_allowed_connections = m_running_transfers;
+        }
+        else
+        {
+            m_allowed_connections = std::size_t(1);
+        }
+    }
+
+    void Mirror::increase_running_transfers()
+    {
+        lock_guard lock(m_stats_mutex);
+        ++m_running_transfers;
+        if (m_max_tried_connections < m_running_transfers)
+        {
+            m_max_tried_connections = m_running_transfers;
+        }
+    }
+
+    void Mirror::update_transfers_done(bool success, bool record_success)
+    {
+        lock_guard lock(m_stats_mutex);
+        --m_running_transfers;
+        if (record_success)
+        {
+            if (success)
+            {
+                ++m_successful_transfers;
+            }
+            else
+            {
+                ++m_failed_transfers;
+            }
+        }
+    }
+
+    /************************************
+     * PassThroughMirror implementation *
+     ************************************/
+
+    PassThroughMirror::PassThroughMirror()
+        : Mirror(PassThroughMirror::id())
+    {
+    }
+
+    MirrorID PassThroughMirror::id()
+    {
+        return MirrorID("");
```

Comment:
Consider making a named constant for `MirrorID("")` and using that instead, to clarify what this means.